### PR TITLE
T13314

### PR DESCRIPTION
**Pull request by tchx84**
https://phabricator.endlessm.com/T13314

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+# Generated files that 'make clean' should erase
+CLEANFILES =
+DISTCLEANFILES =
+EXTRA_DIST =
+
+# # # BUILD RULES # # #
+
+# Launcher script
+CLEANFILES += basin
```
Comment (ptomato) [inline Makefile.am]: This is duplicated on line 50.

File: autogen.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
+FILE_MUST_EXIST=src/main.js
```
Comment (ptomato) [inline autogen.sh]: I prefer to use `configure.ac` for this purpose, since source files are often renamed.

File: Makefile.am [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	src/index.js \
+	src/packer.js \
+	$(NULL)
+
+nobase_pkgdata_DATA = \
+	$(js_sources) \
+	$(NULL)
+
+edit = sed \
+	-e 's|@bindir[@]|$(bindir)|g' \
```
Comment (ptomato) [inline Makefile.am]: Only `@pkgdatadir@` is present in `basin.in`, so you don't need to substitute `@bindir@` and `@datadir@`.

File: configure.ac [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+# Initialization
+# --------------
+# Initialize Autoconf: package name, version, bug report address, tarball name,
+# website
+AC_INIT([Basin], [0],
+    [], [basin], [http://endlessm.com])
+# Initialize Automake: enable all warnings and do not insist on GNU standards
+# no-portability suppresses warnings about syntax specific to GNU make
+AM_INIT_AUTOMAKE([-Wall -Wno-portability foreign 1.11 tar-ustar])
```
Comment (ptomato) [inline configure.ac]: Use `dist-xz dist-no-gzip` as well, for modern packages

File: configure.ac [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+AC_CONFIG_MACRO_DIR([m4])
+
+AC_CACHE_SAVE
+
+# Required build tools
+# --------------------
+# Make sure we can create directory hierarchies
+AC_PROG_MKDIR_P
+# Various tools
+AC_PROG_AWK
```
Comment (ptomato) [inline configure.ac]: I don't think awk is used anywhere

File: configure.ac [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+AC_CACHE_SAVE
+
+# Required libraries
+# ------------------
+EOS_CHECK_GJS_GIR([Gio], [2.0])
```
Comment (ptomato) [inline configure.ac]: Use [`AX_CHECK_GIRS_GJS`](https://www.gnu.org/software/autoconf-archive/ax_check_girs_gjs.html) instead.

File: m4/eos-gir.m4 [hunk 7]
```diff
@@ -0,0 +1,86 @@
+dnl Copyright 2017 Endless Mobile, Inc.
```
Comment (ptomato) [inline m4/eos-gir.m4]: Delete this, use the upstream version instead.

File: src/index.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    finish: function () {
+        let [file, stream] = Gio.File.new_tmp('db_XXXXXX');
+        let out_stream = stream.get_output_stream();
+
+        this._db.commit();
+        this._db.compact_to_fd(out_stream.get_fd(), Xapian.DatabaseCompactFlags.SINGLE_FILE);
+        this._db.close();
+
+        return file;
+    }
```
Comment (ptomato) [inline src/index.js]: nitpick: Use trailing commas everywhere

File: src/index.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+                    'prefix': TAG_PREFIX
+                },
+                {
+                    'field': 'id',
+                    'prefix': ID_PREFIX
+                }
+            ]
+        };
+
+        this._db = new Xapian.WritableDatabase({ path: db_dir,
```
Comment (ptomato) [inline src/index.js]: nitpick: Use our usual indentation style here

File: src/index.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+                                                 action: Xapian.DatabaseAction.CREATE_OR_OVERWRITE,
+                                                 flags: Xapian.DatabaseFlags.BACKEND_GLASS });
+        this._db.init(null);
+        this._db.set_metadata('XbPrefixes', JSON.stringify(prefixes));
+        this._db.set_metadata('XbStopwords', JSON.stringify(stopwords));
```
Comment (ptomato) [inline src/index.js]: There are no stopwords yet, maybe better to leave this out?

File: src/packer.js [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        this._json = JSON.parse(data);
+
+        this._shard = new Shard.Shard(shard_path);
+        this._index = new Index.Index();
+        this._links = new Links.Links();
+    },
+
+    run: function () {
+        print('Writing articles');
+        this._json['articles'].forEach((entry) => {
```
Comment (ptomato) [inline src/packer.js]: optional: `this._json['articles'].forEach(this._add_article, this);` is more concise but arguably less readable

**Review by ptomato — CHANGES_REQUESTED** [review]
Some small changes. I will test this together with the other pull request.

File: src/index.js (hunk 10 above)
Comment (tchx84) [reply src/index.js]: Our back-end seems to expect that field to be present, so I rather leave it there, even if its empty for now.

Comment (ptomato) [reply src/index.js]: That would be good to explain in a comment.

Comment (tchx84) [reply src/index.js]: Sure, let me push that change.

File: autogen.sh [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+    echo "This script will run ./configure automatically. If you wish to pass "
+    echo "any arguments to it, please specify them on the $0 "
+    echo "command line. To disable this behavior, have NOCONFIGURE=1 in your "
+    echo "environment."
+fi
+
+# Run the actual tools to prepare the clean checkout
+mkdir -p m4
+autoreconf -fi -I m4 || exit $?
+rm -f po/Makevars.template
```
Comment (ptomato) [inline autogen.sh]: You don't need this line if not using Gettext.

File: src/index.js [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+        this._db = new Xapian.WritableDatabase({
+            path: db_dir,
+            action: Xapian.DatabaseAction.CREATE_OR_OVERWRITE,
+            flags: Xapian.DatabaseFlags.BACKEND_GLASS,
+        });
+        this._db.init(null);
+        this._db.set_metadata('XbPrefixes', JSON.stringify(prefixes));
+
+        // XXX stopwords field is required so leave it empty for now
```
Comment (ptomato) [inline src/index.js]: No `XXX` needed, this is not really a FIXME situation

File: src/links.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
+const Lang = imports.lang;
+
+const EosShard = imports.gi.EosShard;
+const Gio = imports.gi.Gio;
+
+const Links = new Lang.Class({
+    Name: 'Links',
+
+    _init: function () {
+        this._links = {};
```
Comment (ptomato) [inline src/links.js]: Better to use a `Map` here.

File: src/main.js [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    'basin shard creation tool for Knowledge Apps.',
+].join('\n');
+
+function main () {
+    let argv = ARGV.slice();
+
+    if (argv.length !== 2)
+        fail_with_message(USAGE);
+
+    let json_path = argv[0];
```
Comment (ptomato) [inline src/main.js]: `let [json_path, shard_path] = argv`?

**Review by ptomato — CHANGES_REQUESTED** [review]
Some more small changes, the most important of which is to use a `Map` for the dictionary.

**Review by ptomato — APPROVED** [review]
**CR pass**

**Pull request merged**